### PR TITLE
first go at fixing #1790 (lchown part)

### PR DESCRIPTION
**Pull request by simevo**
TODO:
- move platform-specfic code to the right place
- test
- doc

### CODE REVIEW

File: include/uv.h [hunk 1]
```diff
@@ -1392,6 +1393,12 @@ UV_EXTERN int uv_fs_fchown(uv_loop_t* loop,
                            uv_uid_t uid,
                            uv_gid_t gid,
                            uv_fs_cb cb);
+UV_EXTERN int uv_ls_chown(uv_loop_t* loop,
```
Comment (cjihrig) [inline include/uv.h]: `uv_fs_lchown()`?

File: src/unix/fs.c [hunk 2]
```diff
@@ -1212,6 +1213,20 @@ int uv_fs_fchown(uv_loop_t* loop,
 }
 
 
+int uv_fs_lchown(uv_loop_t* loop,
+                 uv_fs_t* req,
+                 uv_os_fd_t file,
```
Comment (cjihrig) [inline src/unix/fs.c]: The signature doesn't match up.

File: src/unix/fs.c [hunk 3]
```diff
@@ -1212,6 +1213,20 @@ int uv_fs_fchown(uv_loop_t* loop,
...
+                 uv_uid_t uid,
+                 uv_gid_t gid,
+                 uv_fs_cb cb) {
+  INIT(LCHOWN);
+  req->file = file;
```
Comment (cjihrig) [inline src/unix/fs.c]: This doesn't seem correct to me.

**Comment by simevo** [comment]
the existing `chown` and `fchown` functions "_are not implemented on Windows_":
http://docs.libuv.org/en/v1.x/fs.html#c.uv_fs_chown
so neither should be these ones I believe.

there's a TODO in the tests because I have though yet about how to check that `lchown` really does not change the linked-to file

also:
```
not ok 45 - fs_chown
# exit code 6
# Output from process `fs_chown`:
# Assertion failed in ../test/test-fs.c on line 1561: r == 0
```

**Comment by cjihrig** [comment]
> the existing chown and fchown functions "are not implemented on Windows":

There are still stubs in the code - https://github.com/libuv/libuv/blob/ff167ea5db66903c1a2f65b29ea0147b1108f2c9/src/win/fs.c#L2258.

**Comment by simevo** [comment]
Now it seems to compile and generate the library fine on windows with VS 2015.

There's still the test failing and the missing test (see TODO in `test/test-fs.c`).
I wonder what user id -1 stands for, on my system:
```
grep UID /etc/login.defs 
UID_MIN                  1000
UID_MAX                 60000
...
```

**Comment by simevo** [comment]
Hi what else can I do to get this PR into libuv ? Thanks

File: include/uv.h [hunk 4]
```diff
@@ -1392,6 +1393,12 @@ UV_EXTERN int uv_fs_fchown(uv_loop_t* loop,
                            uv_uid_t uid,
                            uv_gid_t gid,
                            uv_fs_cb cb);
+UV_EXTERN int uv_fs_lchown(uv_loop_t* loop,
+                          uv_fs_t* req,
```
Comment (cjihrig) [inline include/uv.h]: Can you line all of these arguments up with the `uv_loop_t` on the previous line.

File: src/unix/fs.c [hunk 5]
```diff
@@ -1212,6 +1213,20 @@ int uv_fs_fchown(uv_loop_t* loop,
 }
 
 
+int uv_fs_lchown(uv_loop_t* loop,
+                uv_fs_t* req,
```
Comment (cjihrig) [inline src/unix/fs.c]: Please line the arguments up.

File: test/test-fs.c [hunk 6]
```diff
@@ -84,6 +84,7 @@ static int chmod_cb_count;
 static int fchmod_cb_count;
 static int chown_cb_count;
 static int fchown_cb_count;
+static int lchown_cb_count;
```
Comment (cjihrig) [inline test/test-fs.c]: This variable isn't asserted on anywhere.

File: test/test-fs.c [hunk 7]
```diff
@@ -243,6 +244,13 @@ static void chown_cb(uv_fs_t* req) {
   uv_fs_req_cleanup(req);
 }
 
+static void lchown_cb(uv_fs_t* req) {
```
Comment (cjihrig) [inline test/test-fs.c]: This isn't called.

**Comment by cjihrig** [comment]
@bnoordhuis do you have the historical context on why these functions just return 0 on Windows instead of `ENOSYS`?

**Comment by cjihrig** [comment]
Sorry, "these functions" refers to the chown family of functions.

**Comment by bnoordhuis** [comment]
It's because there is no meaningful way to implement them, user/group/other isn't a thing on Windows.

I don't think it was a deliberate decision to quietly ignore instead of returning ENOSYS, just something that was deemed Good Enough at the time and made node's test suite pass. :-)

**Comment by cjihrig** [comment]
> I don't think it was a deliberate decision to quietly ignore instead of returning ENOSYS, just something that was deemed Good Enough at the time and made node's test suite pass.

That's what I was looking for. Thanks!

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/868/

**Comment by simevo** [comment]
all CI failures seem unrelated to me:

https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=debian7-docker-armv7/

```
ERROR: Error cloning remote repo 'origin'
```

https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=alpine-last-latest-x64/
https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=alpine-latest-x64/
```
not ok 20 - connect_unspecified
# exit code 134
# Output from process `connect_unspecified`:
# Assertion failed in test/test-connect-unspecified.c on line 56: uv_tcp_connect(&connect6, &socket6, (const struct sockaddr*) &addr6, connect_6) == 0
```

https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=centos6-64/
https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=centos6-64-gcc44/
https://ci.nodejs.org/job/libuv-test-commit-linux/910/nodes=ubuntu1404-gyp-64/

```
not ok 103 - getaddrinfo_fail_sync
# timeout
# Output from process `getaddrinfo_fail_sync`: (no output)
```

https://ci.nodejs.org/job/libuv-test-commit-aix/879/nodes=aix61-ppc64/

```
not ok 250 - tcp_ping_pong
# exit code 393222
# Output from process `tcp_ping_pong`: (no output)
Output from process `tcp_ping_pong`:
# Assertion failed in test/test-ping-pong.c on line 144: status == 0
...
not ok 256 - tcp_ref3
# exit code 393222
# Output from process `tcp_ref3`: (no output)
Output from process `tcp_ref3`:
# Assertion failed in test/test-ref.c on line 96: status == 0
...
not ok 262 - tcp_write_fail
# exit code 393222
# Output from process `tcp_write_fail`: (no output)
Output from process `tcp_write_fail`:
# Assertion failed in test/test-tcp-write-fail.c on line 77: status == 0
```

https://ci.nodejs.org/job/libuv-test-commit-zos/724/nodes=zos-s390x/
```
not ok 309 - udp_multicast_join6
# exit code 3
# Output from process `udp_multicast_join6`:
# Assertion failed in ../test/test-udp-multicast-join6.c on line 136: r == 0
# CEE5207E The signal SIGABRT was received.
```

**Comment by cjihrig** [comment]
Landed on master in https://github.com/libuv/libuv/commit/33b6db3909dca45fa543ab6405b934f1333fd213. Thanks!

I'm going to look into backporting this to v1.x as well.

**Comment by cjihrig** [comment]
Landed on v1.x in https://github.com/libuv/libuv/commit/aa28f7d5bf36b4df1e606ed47364d91d896b1f8b.

Here is a CI run of that commit with no related failures: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/940/

**Pull request closed without merging**